test/e2e/clusterctl_upgrade.go

```diff
@@ -307,6 +318,19 @@ func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpg
 			input.PostUpgrade(managementClusterProxy)
 		}
 
+		// After the upgrade make check that there were no unexpected rollouts.
+		Consistently(func() bool {
+			postUpgradeMachineList := &clusterv1.MachineList{}
+			err = managementClusterProxy.GetClient().List(
+				ctx,
+				postUpgradeMachineList,
+				client.InNamespace(testNamespace.Name),
+				client.MatchingLabels{clusterv1.ClusterLabelName: workLoadClusterName},
+			)
+			Expect(err).NotTo(HaveOccurred())
+			return machinesMatch(preUpgradeMachineList, postUpgradeMachineList)
+		}, input.E2EConfig.GetIntervals(specName, "check-upgrade-no-rollout")...).Should(BeTrue())
```

Comment:
Do I see correctly that this checks every 30 sec for 10 minutes if something changes?

Would it be enough to do this for maybe 30 sec or a minute?

Not sure if we want to / have to make this configurable. This should be mainly a matter of how fast the machine controller would trigger a rollout and this should be the same for all providers

Reply:
Can we add a message to "Should"?

Reply:
>Do I see correctly that this checks every 30 sec for 10 minutes if something changes?

Yes, I started with 10m because I believe it is the default sync time and it thought "this should be enough to make sure atleast one reconcile loop runs and trigger changes if any". No strong opinions on using a different time.

>Would it be enough to do this for maybe 30 sec or a minute?

1 minutes seems to short. With some local testing I observed that in quite a few times it takes more than a minute for KCP to start a rollout. We might increase our chances of catching any rollouts with such a short check duration.

How about 5 minutes?

>Not sure if we want to / have to make this configurable. This should be mainly a matter of how fast the machine controller would trigger a rollout and this should be the same for all providers

Just to note, rollouts could also be triggered for control plane providers (KCP) and since technically people can bring in their own control plane providers which might act differently I originally thought it might be better to make this configurable. No strong opinions on keeping it this way.

>Can we add a message to "Should"?

done.

Reply:
The problem with 10m is that the test now takes 10m longer. If we can think of any way to avoid this, that would be great. 

I would have assumed that KCP immediately rollouts if it does, because the controller is upgraded and on start it does a full list of KubeadmControlPlanes.

Reply:
It would be nice to know why it takes so long for the rollout to start. I would have assumed it starts almost immediately once the KCP controller is up. But maybe we don't wait in `UpgradeManagementClusterAndWait` until the controller is up? (I don't remember how this is done by clusterctl, maybe there is room for improvement there)

Reply:
Some findings: When the new controller spin up the the reconciler and watchers and all take around 1 minute to setup. 
Running the test several times I observed that we tend to observe the changes in under 3 minutes in all cases. With this I think we can drop the duration from 10m to 3m. 
Thoughts?

Reply:
Sounds like a good improvement. 

I wonder if clusterctl upgrade currently waits until the new controllers come up (i.e. if they are ready == webhooks are up).

 If it doesn't maybe it should? So that at the end of clusterctl upgrade you actually now if it worked. This would be an independent improvement though.

Reply:
>I wonder if clusterctl upgrade currently waits until the new controllers come up (i.e. if they are ready == webhooks are up).
>
>If it doesn't maybe it should? So that at the end of clusterctl upgrade you actually now if it worked. This would be an independent improvement though.

It does not wait for the deployments. It just creates the new deployments and exits. It might be a good UX to wait and error if any deployment does not come up. We do something similar for init. It should be addressed in a separate PR.

>Or if we don't want it in clusterctl, maybe the e2e test should wait / verify that the controllers are up? (in case it's not already happening)

The tests do wait for the deployments to be ready after the upgrade is performed. It waits for the `Available` condition on the deployment to be true.

The 1m for the initial setup happens after the controller code starts running (deployment is up and pods are running).  During the tests I ran I observed that we always detect it under (sometimes at) 2.5 mins (includes the 1min for the setup) after the deployments are up and running (we already wait for them in the upgrade step). To be safe I bumped it up to 3m and this seems to be safe enough to always detect any rollouts.